test/functional/dnsmasq_controller_test.go

```diff
@@ -0,0 +1,211 @@
+			Expect(instance.Status.ReadyCount).To(Equal(int32(0)))
+			Expect(instance.Spec.ContainerImage).Should(Equal(containerImage))
+		})
+
+		It("creates service account, role and rolebindig", func() {
+			th.ExpectCondition(
+				dnsMasqName,
+				ConditionGetterFunc(DNSMasqConditionGetter),
+				condition.ServiceAccountReadyCondition,
+				corev1.ConditionTrue,
+			)
+			sa := th.GetServiceAccount(dnsMasqServiceAccountName)
+
+			th.ExpectCondition(
+				dnsMasqName,
+				ConditionGetterFunc(DNSMasqConditionGetter),
+				condition.RoleReadyCondition,
+				corev1.ConditionTrue,
+			)
+			role := th.GetRole(dnsMasqRoleName)
+			Expect(role.Rules).To(HaveLen(2))
+			Expect(role.Rules[0].Resources).To(Equal([]string{"securitycontextconstraints"}))
+			Expect(role.Rules[1].Resources).To(Equal([]string{"pods"}))
+
+			th.ExpectCondition(
+				dnsMasqName,
+				ConditionGetterFunc(DNSMasqConditionGetter),
+				condition.RoleBindingReadyCondition,
+				corev1.ConditionTrue,
+			)
+			binding := th.GetRoleBinding(dnsMasqRoleBindingName)
+			Expect(binding.RoleRef.Name).To(Equal(role.Name))
+			Expect(binding.Subjects).To(HaveLen(1))
+			Expect(binding.Subjects[0].Name).To(Equal(sa.Name))
+		})
+
+		It("generated a ConfigMap holding dnsmasq key=>value config options", func() {
+			th.ExpectCondition(
+				dnsMasqName,
+				ConditionGetterFunc(DNSMasqConditionGetter),
+				condition.ServiceConfigReadyCondition,
+				corev1.ConditionTrue,
+			)
+
+			configData := th.GetConfigMap(dnsMasqName)
+			Expect(configData).ShouldNot(BeNil())
+			Expect(configData.Data[dnsMasqName.Name]).Should(
+				ContainSubstring("server=1.1.1.1"))
+			Expect(configData.Labels["dnsmasq.openstack.org/name"]).To(Equal(dnsMasqName.Name))
+		})
+
+		It("stored the input hash in the Status", func() {
+			Eventually(func(g Gomega) {
+				instance := GetDNSMasq(dnsMasqName)
+				g.Expect(instance.Status.Hash).Should(HaveKeyWithValue("input", Not(BeEmpty())))
+			}, timeout, interval).Should(Succeed())
+		})
+
+		It("exposes the service", func() {
+			th.ExpectCondition(
+				dnsMasqName,
+				ConditionGetterFunc(DNSMasqConditionGetter),
+				condition.ExposeServiceReadyCondition,
+				corev1.ConditionTrue,
+			)
+
+			svc := th.GetService(types.NamespacedName{
+				Namespace: namespace,
+				Name:      fmt.Sprintf("dnsmasq-%s-ctlplane", dnsMasqName.Name)})
+			Expect(svc.Labels["service"]).To(Equal("dnsmasq"))
+		})
+
+		It("creates a Deployment for the service", func() {
+
+			th.ExpectConditionWithDetails(
+				dnsMasqName,
+				ConditionGetterFunc(DNSMasqConditionGetter),
+				condition.DeploymentReadyCondition,
+				corev1.ConditionFalse,
+				condition.RequestedReason,
+				condition.DeploymentReadyRunningMessage,
+			)
+
+			deploymentName = types.NamespacedName{
+				Name:      fmt.Sprintf("dnsmasq-%s", dnsMasqName.Name),
+				Namespace: namespace,
+			}
+
+			time.Sleep(1000 * time.Millisecond)
+
+			depl := th.GetDeployment(deploymentName)
```

Comment:
In general you can use Eventually to assert async conditions. GetDeployment already uses Eventually to wait for the Deployment to exists. So if the only thing the sleep was needed is to wait for the Deployment to be created then you can simply remove the sleep and the Eventually block in GetDeployment will wait for it.

In my local testing this is the case, i.e. I removed the sleep and the test still passes. 

But if for some reason the reconcile would first create the Deployment and then later it would update it (e.g. setting the Replicas) then GetDeployment might return right after the Deployment is created but before it was updated in a later reconcile. If you have this situation then you can use an Eventually block like:
```golang
Eventually(func(g Gomega){
    depl := th.GetDeployment(deploymentName)
    g.Expect(int(*depl.Spec.Replicas)).To(Equal(1))
}, timeout, interval).Should(Succeed())
```
This Eventually could then waits for both conditions: i) the GetDeployment returns a deployment so it exsists, ii) the Replicas in that deployment in equal to 1.